libvast/src/experimental_table_slice_builder.cpp

```diff
@@ -809,12 +873,50 @@ type make_vast_type(const arrow::DataType& arrow_type) {
   return caf::visit(f, arrow_type);
 }
 
+type enrich_type_with_metadata(
+  const type& t,
+  const std::shared_ptr<const arrow::KeyValueMetadata>& metadata) {
+  if (!metadata)
+    return t;
+  if (auto ml = metadata->Get("VAST:max_level"); !ml.ok()) {
+    VAST_INFO("skipping non-VAST metadata for field '{}'. Metadata was: '{}'",
+              t.name(), metadata->ToString());
+    return t;
+  } else {
+    auto enriched_type = t;
+    int max_level = stoi(*ml);
+    // working from innermost to outermost nesting
+    for (int l = max_level; l >= 0; --l) {
+      auto name = metadata->Get(fmt::format("VAST:name:{}", l)).ValueOr("");
```

Comment:
We call `metadata->Get(...)` a total of `1 + 2 * max_level` times. Do you think it's worth switching this to an approach that simply iterates over the metadata instead?

E.g., something like this:

```
auto max_level = metadata->Get(...);
if (max_level == 0)
  return t;
auto names_and_attributes = std::vector<std::pair<std::string_view, struct type::attribute>>{};
names_and_attributes.resize(max_level);
for (const auto& [key, value] : *metadata) {
  if (!value.starts_with("VAST:"))
    continue;
  // parse key and enter into names_and_attributes at parsed index
}
// iterate in reverse to reconstruct type
for (const auto* it = names_and_attributes.rbegin(); it != names_and_attributes.rend(); ++it) {
  t = type{it->first, t, it->second};
}
```

Not 100% sure if we should do this, but the current approach is very allocation heavy with all the formatting going on that we can mostly avoid.

Reply:
Good point. I'll look into this a bit. The `Get` is implemented as a linear scan, so this might indeed not be optimal, and iterating over the metadata entries instead is definitely an improvement.

Reply:
We paired and came up with a more efficient solution along the lines of the code above, using our `parseable` machinery to reduce the allocations and iterating over the metadata only once to avoid the `O(n^2)` that result from the repeated calls to `metadata->Get()`.

Reply:
Revisiting this topic: now that we sequentially scan all metadata fields anyway, the need for `VAST:nesting-depth` is no longer there. 
I think we can get rid of that, which is also more closely what we initially thought of in our design document.